specs/simple-serialize.md

```diff
@@ -449,6 +449,77 @@ assert item_index == start + LENGTH_BYTES + length
 return typ(**values), item_index
 ```
 
+### Tree hash
+
+The below algorithm is defined recursively in the case of lists and containers, and it outputs a value equal to or less than 32 bytes in size. For the final output only (ie. not intermediate outputs), if the output is less than 32 bytes, right-zero-pad it to 32 bytes. The goal is collision resistance *within* each type, not between types.
+
+We define `hash(x)` as `BLAKE2b-512(x)[0:32]`.
+
+#### uint: 8/16/24/32/64/256, bool, address, hash32
+
+Return the serialization of the value.
+
+#### bytes, hash96
+
+Return the hash of the serialization of the value.
+
+#### List/Vectors
+
+First, we define some helpers and then the Merkle tree function. The constant `CHUNK_SIZE` is set to 128.
+
+```python
+# Returns the smallest power of 2 equal to or higher than x
+def next_power_of_2(x):
+    return x if x == 1 else next_power_of_2((x+1) // 2) * 2
+
+# Extends data length to a power of 2 by minimally right-zero-padding
+def extend_to_power_of_2(data):
+    return data + b'\x00' * (next_power_of_2(len(data)) - len(data))
+
+# Concatenate a list of homogeneous objects into data and pad it
+def list_to_glob(lst):
+    if len(lst[0]) != next_power_of_2(len(lst[0])):
```

Comment:
Why are we just checking the 0th item? I don't think there is any guarantee that the data length of these objects are the same considering they can have variable length lists inside of them.

Reply:
Oh is it because the items have already been `tree_hash`ed to get to this point?

Reply:
Why do `extend_to_power_of_2` at all?

Reply:
> In my understanding, it guarantees that despite the object's length, chunks will fit boundaries of extended objects. A list of address objects is a good example here.

Yes, exactly this.

> Oh is it because the items have already been tree_hashed to get to this point?

Yep! If the subtype is anything variable-sized, then the tree hash will have 32 bytes.

Reply:
> chunks will fit boundaries of extended objects

alternative that packs 6 instead of 4 addresses (assuming len < CHUNKSIZE as simplicifcation), by padding only once per chunk:

```python
items_per_chunk = CHUNKSIZE // len(l[0])
padding = b'\00' * (CHUNKSIZE - items_per_chunk * len(l[0]))
chunkz = [b''.join(l[i:i+items_per_chunk])+padding for i in range(0, len(l), items_per_chunk)]
```